onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -43,73 +45,142 @@ using namespace onnxruntime::common;
 */
+
 template <typename C, typename T = typename C::value_type>
 static std::string ConcatNames(
-    C const& items, std::string (*f)(const T& n) = [](const T& n) { return n->Name(); }) {
-  std::vector<std::string> names;
-  std::transform(items.begin(), items.end(), back_inserter(names), f);
-  return std::accumulate(names.begin(), names.end(), std::string(), [](const std::string& a, const std::string& b) { return a + ", " + b; });
+    const C& items, std::string (*f)(const T& n) = [](const T& n) { return n->Name(); }) {
+  std::ostringstream os;
+  if (!items.empty()) {
+    auto first = items.begin();
+    for (size_t i = 0, limit = items.size() - 1; i < limit; ++i, ++first) {
+      os << f(*first) << ", ";
+    }
+    os << f(*first);
+  }
+  return os.str();
 }
 
 // GetName
-// Collect the node-arg names and the correspnding consumers name for reporting
-static std::string GetName(const std::pair<const NodeArg*, std::vector<Node*>>& p) {
-  return p.first->Name() + " feeding " + ConcatNames(p.second) + "; ";
+// Collect the node-arg names and the corresponding consumers name for reporting
+static std::string GetName(const NodeArgToConsumerMap::value_type& p) {
+  std::string result(p.first->Name());
+  result.append(" feeding ").append(ConcatNames(p.second)).append("; ");
+  return result;
 };
 /*
 * The collection fp16_allow_ops, specifies for a given propagate_cast_ops level, a vector of node op_types that
 * the code is allowed to propage Cast operations across. The user may specify a custom list of optypes using level 0.
 * The opcodes are split into multiple levels. Cast propagation is done based on the level. Level 2 op code
 * list includes Level 1 list also.
 */
-static std::vector<std::unordered_set<std::string>> fp16_allow_ops = {
-    /* Level 0 */ {},
-    /* Level 1 */ {"Expand", "Transpose", "Relu", "Reshape", "Split", "Tanh", "Squeeze", "Unsqueeze"},
-    /* Level 2 */ {"Add", "BiasGelu", "Dropout", "FastGelu", "Gather", "Gelu", "LayerNormalization", "Where"}};
+static const InlinedHashSet<std::string_view> allow_list_level1 =
+    {"Expand", "Transpose", "Relu", "Reshape", "Split", "Tanh", "Squeeze", "Unsqueeze"};
+static const InlinedHashSet<std::string_view> allow_list_level2 =
+    {"Add", "BiasGelu", "Dropout", "FastGelu", "Gather", "Gelu", "LayerNormalization", "Where"};
 
 /*
 *  The following two maps specify the opcode to input and opcode to output mappings to list the inputs/outputs to consider while propagating
 *  cast operations. All other inputs/outputs not listed in this table are not relevant for deciding whether an operation
 *  performed in float or float16. If an opcode is not listed in these tables, the code will look at all the inputs and outputs to validate 
 *  transformation.
 */
-static std::unordered_map<std::string, std::vector<int>> opcode_to_input_map = {
+static const InlinedHashMap<std::string_view, std::array<int, 3>> opcode_to_input_map = {
```

Comment:
perhaps `InlinedHashMap<std::string_view, InlinedVector<int>>`? otherwise `{"Gather", {0}}` results in a `std::array<int, 3>` value of {0, 0, 0}